### PR TITLE
bpo-29612 : Fix hard link creation in tar extraction when it points to themselves

### PR DESCRIPTION
**Pull request by j0ack**
When creating a link from a tarball, if the link target and the file has the same path it raised an exception.

We simply ignore this case at the beginning of the method.

https://bugs.python.org/issue29612

### CODE REVIEW

File: Lib/tarfile.py [hunk 1]
```diff
@@ -2272,6 +2272,13 @@ def makelink(self, tarinfo, targetpath):
           (platform limitation), we try to make a copy of the referenced file
           instead of a link.
         """
+        # do not create link if target path is equals to link target
+        if (
```
Comment (matrixise) [inline Lib/tarfile.py]: could you use the PEP8 style for the condition? 
for example
```python
if hasattr(tarinfo, "_link_target") and tarinfo._link_target == targetpath:
```

**Review by matrixise — CHANGES_REQUESTED** [review]
review the style for the condition

File: Lib/tarfile.py (hunk 1 above)
Comment (j0ack) [reply Lib/tarfile.py]: The line is 83 characters long if I do not split it into two lines.

**Comment by j0ack** [comment]
done

When I do not split the condition into two lines, the result is 83 characters long, which is not pep8 right ?

**Comment by matrixise** [comment]
@j0ack not necessarily, the max length can be between 80 and 100, check the PEP8 https://www.python.org/dev/peps/pep-0008/#maximum-line-length

**Comment by encukou** [comment]
That document says:

> The Python standard library is conservative and requires limiting lines to 79 characters (and docstrings/comments to 72).

So, yes, 83 characters is too long.

File: Lib/tarfile.py (hunk 1 above)
Comment (Mariatta) [reply Lib/tarfile.py]: The preferred way to split it is as follows:
```
if hasattr(tarinfo, "_link_target") and \
    tarinfo._link_target == targetpath:
```

**Comment by j0ack** [comment]
That's the reason why I splitted my condition into two lines. 

Let me know if I need to revert to my previous commit

**Comment by j0ack** [comment]
I can reduce the line length with `getattr` default parameter

**Comment by JulienPalard** [comment]
@j0ack will try to commit this to master first then backport it to 2.7.

**Comment by j0ack** [comment]
See #8700

**Pull request closed without merging**